### PR TITLE
Initial version of MCTS for AlphaZero

### PR DESCRIPTION
**Pull request by alejandromarcu**
AlphaZero uses a modified version of MCTS.   I just copied the existing implementation and modified it.  There are many changes spread out across the class, so it seems that trying to re-use it would be tricky (I know OpenSpiel does it, but haven't checked how they do it).
I haven't tested it yet, but I wanted to send what I did so far in case anyone else is interested in working with this this week, since I won't have a lot of time.

### CODE REVIEW

File: deep_quoridor/src/agents/alphazero.py [hunk 1]
```diff
@@ -30,6 +37,119 @@ class AlphaZeroParams(SubargsBase):
     # Exploration vs exploitation.  0 is pure exploitation, infinite is random exploration.
     temperature: float = 1.0
 
+    # Number of searches
+    n: int = 1000
+    # A higher number favors exploration over exploitation
+    c: float = 1.4
+
+
+class AzNode:
```
Comment (adamantivm) [inline deep_quoridor/src/agents/alphazero.py:46]: :donkey:

File: deep_quoridor/src/agents/alphazero.py [hunk 2]
```diff
@@ -30,6 +37,119 @@ class AlphaZeroParams(SubargsBase):
...
+            # Traverse down the tree guided by maximum UCB until we find a node to expand
+            node = self.select(root)
+
+            if node.game.check_win(good_player):
+                value = 1
+            elif node.game.check_win(1 - good_player):
+                value = -1
+            else:
+                with torch.no_grad():
+                    policy, value = self.nn(node.game)
```
Comment (adamantivm) [inline deep_quoridor/src/agents/alphazero.py:135]: value is not used during evaluation, right?

File: deep_quoridor/src/agents/alphazero.py [hunk 3]
```diff
@@ -30,6 +37,119 @@ class AlphaZeroParams(SubargsBase):
...
+        for _ in range(self.params.n):
+            # Traverse down the tree guided by maximum UCB until we find a node to expand
+            node = self.select(root)
+
+            if node.game.check_win(good_player):
+                value = 1
+            elif node.game.check_win(1 - good_player):
+                value = -1
+            else:
+                with torch.no_grad():
```
Comment (adamantivm) [inline deep_quoridor/src/agents/alphazero.py:134]: I think the way OpenSpiel generalizes the MCTS implementation is by plugging in an "Evaluator" here, which returns what the thing needs to expand and back-propagate. 

So in the default version the "evaluator" is the simulation you already implmeented. And in this other verison, the evaluator is the NN evaluator.

I don't think it's necessary to do it in that way, I just mention it in case it's useful to compare the implementaiton to confirm details and clear doubts as we go with our implementation.

File: deep_quoridor/src/agents/alphazero.py [hunk 4]
```diff
@@ -30,6 +37,119 @@ class AlphaZeroParams(SubargsBase):
...
+
+        self.ucb_c = ucb_c
+        self.prior = prior
+
+    def should_expand(self):
+        return len(self.children) == 0
+
+    def expand(self, policy_probs: np.ndarray):
+        """
+        Create a child of the current node.
```
Comment (adamantivm) [inline deep_quoridor/src/agents/alphazero.py:71]: Here you create all the childs, right? Maybe update the comment to match?

Also consider if you wish, just to keep both evolving in parallel to make them easier to compare, to also update mcts.py to do it in this same way to match. If I understand things correctly, the default behavior would be to expand all with equal probability, maybe? I am probably missing something. Just trying to use the opportunity to understand a bit more, that's all.

**Review by adamantivm — APPROVED** [review]
Awesome to see progress with this, cucu!

I left some comments but I realize this is just incremental work, feel free to merge if you want and continue implementaiton as normal.

File: deep_quoridor/src/agents/alphazero.py [hunk 5]
```diff
@@ -30,6 +37,119 @@ class AlphaZeroParams(SubargsBase):
...
+        self.value_sum += value
+        self.visit_count += 1
+
+        if self.parent is not None:
+            self.parent.backpropagate(-value)
+
+
+class AzMCTS:
+    def __init__(self, nn: nn.Module, params: AlphaZeroParams):
+        self.nn = nn
```
Comment (muralx) [inline deep_quoridor/src/agents/alphazero.py:108]: something to keep in mind. If you use an NN with Dropout (or the like), you might need to use eval() mode when not training or in the instance that is not training.

File: deep_quoridor/src/agents/alphazero.py (hunk 1 above)
Comment (alejandromarcu) [reply deep_quoridor/src/agents/alphazero.py:46]: Didn't notice, I'm loosing my wittiness.  BTW, at Facebook I once called a class HubsExactIndex, and a while later someone noticed it could be HubSexActIndex 🤣 .  I think linters should start using AI tools to check for those cases and give you a warning

File: deep_quoridor/src/agents/alphazero.py (hunk 2 above)
Comment (alejandromarcu) [reply deep_quoridor/src/agents/alphazero.py:135]: It's used in the backpropagation below.  It's hard to read because how the diff is presented

File: deep_quoridor/src/agents/alphazero.py (hunk 3 above)
Comment (alejandromarcu) [reply deep_quoridor/src/agents/alphazero.py:134]: Yeah, the thing is that there are a few more changes spread out across the classes, which would be annoying to have ifs everywhere

File: deep_quoridor/src/agents/alphazero.py (hunk 4 above)
Comment (alejandromarcu) [reply deep_quoridor/src/agents/alphazero.py:71]: We could change the other MCTS the same way to expand all the nodes (although besides code re-using, I don't think it provides any advantage).   In that one,  it doesn't use priors, so it's an uniform probability.  Basically the advantage here is that it uses the "intuition" of the NN to start exploring the nodes that it estimates are more likely to get to a win.
I'll add a TODO for myself to do the update in the other one

Comment (alejandromarcu) [inline deep_quoridor/src/agents/alphazero.py:71]:
```suggestion
        Create all the children of the current node.
```
